firewood/src/merkle.rs

```diff
@@ -1175,6 +1178,285 @@ impl<S: ShaleStore<Node> + Send + Sync> Merkle<S> {
+                    NodeType::Extension(_) => todo!(),
+                };
+
+                self.key_state = IteratorState::Iterating { last_node, parents };
+
+                return Poll::Ready(Some(Ok(returned_key_value)));
+            }
+            IteratorState::Iterating {
+                last_node,
+                mut parents,
+            } => {
+                match last_node.inner() {
+                    NodeType::Branch(branch) => {
+                        // previously rendered the value from a branch node, so walk down to the first available child
+                        if let Some(child_position) =
+                            branch.children.iter().position(|&addr| addr.is_some())
+                        {
+                            let child_address = branch.children[child_position].unwrap();
+
+                            parents.push((last_node, child_position as u8)); // remember where we walked down from
+
+                            let current_node = self
+                                .merkle
+                                .get_node(child_address)
+                                .map_err(|e| api::Error::InternalError(e.into()))?;
+
+                            let found_key = key_from_parents(&parents);
+
+                            self.key_state = IteratorState::Iterating {
+                                // continue iterating from here
+                                last_node: current_node,
+                                parents,
+                            };
+
+                            found_key
+                        } else {
+                            // Branch node with no children?
+                            return Poll::Ready(Some(Err(api::Error::InternalError(Box::new(
+                                MerkleError::ParentLeafBranch,
+                            )))));
+                        }
+                    }
+                    NodeType::Leaf(leaf) => {
+                        let mut next = parents.pop();
+                        loop {
+                            match next {
+                                None => return Poll::Ready(None),
+                                Some((parent, child_position)) => {
+                                    // Assume all parents are branch nodes
+                                    let children = parent.inner().as_branch().unwrap().chd();
+
+                                    // we use wrapping_add here because the value might be u8::MAX indicating that
+                                    // we want to go down branch
+                                    let mut child_position = child_position.wrapping_add(1);
+
+                                    if let Some(found_offset) = children[child_position as usize..]
+                                        .iter()
+                                        .position(|&addr| addr.is_some())
+                                    {
```

Comment:
```rust
let Some(found_offset) = ... else {
```
^^ this syntax would be nicer here. However, if you prefer `if let Some(..) = expression { `, please save `expression` into a variable instead of allowing it to span multiple lines.

Reply:
I wasn't 100% sure what you were asking for here.  [This commit](https://github.com/ava-labs/firewood/pull/346/commits/dcfb3386858fba6dbd2a73b0470f76f7dedd21cf) was my interpretation. I'm not sure it's any better.